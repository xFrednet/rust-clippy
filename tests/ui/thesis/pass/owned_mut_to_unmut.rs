//@rustc-env: CLIPPY_PETS_PRINT=1
//@rustc-env: CLIPPY_PRINT_MIR=1

#[warn(clippy::borrow_pats)]
fn mut_move_to_immut() {
    let mut x = "Hello World".to_string();
    x.push('x');
    x.clear();
    let x2 = x;
    let _ = x2.len();
}

#[warn(clippy::borrow_pats)]
fn init_with_mut_block() {
    let x2 = {
        let mut x = "Hello World".to_string();
        x.push('x');
        x.clear();
        x
    };
    let _ = x2.len();
}

#[warn(clippy::borrow_pats)]
fn mut_copy_to_immut_shadow() {
    let mut counter = 1;
    counter += 10;

    let counter = counter;

    let _ = counter;
}

#[warn(clippy::borrow_pats)]
fn mut_copy_to_immut() {
    let mut counter = 1;
    counter += 10;

    let snapshot = counter;

    let _ = snapshot;
}

#[warn(clippy::borrow_pats)]
fn mut_copy_to_immut_and_use_after() {
    let mut counter = 1;
    counter += 10;

    let snapshot = counter;

    counter += 3;

    let _ = snapshot + counter;
}

#[warn(clippy::borrow_pats)]
fn main() {
    let mut s = String::new();
    s += "Hey";

    let _ = s;
}
