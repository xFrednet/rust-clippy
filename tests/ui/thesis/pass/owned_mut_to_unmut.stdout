# "mut_move_to_immut"
- x              : (Mutable  , Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Moved, MovedToVar, Borrow, ArgBorrowMut, ModMutShadowUnmut}
- x2             : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasTempBorrow, HasTempBorrowMut}

# "init_with_mut_block"
- x2             : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow}
- x              : (Mutable  , Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Moved, MovedToVar, Borrow, ArgBorrowMut, ModMutShadowUnmut}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasTempBorrow, HasTempBorrowMut}

# "mut_copy_to_immut_shadow"
- counter        : (Mutable  , Owned    , Local   , Copy   , NonDrop , Primitive) {CopiedToVar, Overwrite, ModMutShadowUnmut}
- counter        : (Immutable, Owned    , Local   , Copy   , NonDrop , Primitive) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

# "mut_copy_to_immut"
- counter        : (Mutable  , Owned    , Local   , Copy   , NonDrop , Primitive) {CopiedToVar, Overwrite}
- snapshot       : (Immutable, Owned    , Local   , Copy   , NonDrop , Primitive) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

# "mut_copy_to_immut_and_use_after"
- counter        : (Mutable  , Owned    , Local   , Copy   , NonDrop , Primitive) {CopiedToVar, Overwrite}
- snapshot       : (Immutable, Owned    , Local   , Copy   , NonDrop , Primitive) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

# "main"
- s              : (Mutable  , Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrowMut}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasTempBorrowMut}

